### PR TITLE
Tweak more

### PR DESCRIPTION
**Pull request by dkaukov**
The main problem was that we were violating original contract of otj-server. It was scanning connectors in the order of definitions in the "ot.httpserver.active-connectors" property, and if port is -1 (by default or provided) it allocated PORTx. So "boot" and "default-http" not necessarily first ones. Also "boot" and "default-http" could co-exist together. so, what I did:

1) Pre-build map of PortSelection completely scanning "ot.httpserver.active-connectors"
2) dropped PortIterator
3) portIndex is now inside PortSelector
4) Logic of PortSelector::get() slightly change
  a) First we read spring property and only if it is null or -1 we falling back (we do not want to increment counter, if value provided)
  b) For k8s, to named ports
  c) for singularity, to the ordinal ports (incrementing portIndex)
5) SpringPortSelectionPostProcessor is now simplified and providing everything to the EmbeddedJetty

As a bonus, it is printing "Port Selections" for named connectors as-well

### CODE REVIEW

File: otj-server-core/src/main/java/com/opentable/server/PortSelector.java [hunk 1]
```diff
@@ -92,80 +101,62 @@ public PortSelector(Environment environment) {
...
-                getWithDefault(HTTPSERVER_CONNECTOR_DEFAULT_HTTP_PORT, 0, namedPort, -1));
-
-        // jmx
-        portSelectionMap.put(JMX_PORT, getJMXPort());
-        // actuator
-        //TODO: discuss with dmitry: shouldn't the default be null? eg not flipping this on at all
-        portSelectionMap.put(MANAGEMENT_SERVER_PORT, getActuatorPort());
-
-        return portSelectionMap;
+        Map<String, PortSelection> res = Arrays.stream(environment.getProperty("ot.httpserver.active-connectors", "default-http").split(","))
```
Comment (dkaukov) [inline otj-server-core/src/main/java/com/opentable/server/PortSelector.java:145]: scan active connectors and build full list

File: otj-server-core/src/main/java/com/opentable/server/PortSelector.java [hunk 2]
```diff
@@ -92,80 +101,62 @@ public PortSelector(Environment environment) {
...
-        // First the named port - hence on k8s this gets precedence
-        PortSelection portSelection = get(portName, PortSource.FROM_PORT_NAMED);
-        if (!portSelection.hasValue()) {
-            // Then the "official" spring property
-            portSelection =  get(springPropertyName, PortSource.FROM_SPRING_PROPERTY);
-            if (!portSelection.hasValue()) {
-                // Finally the PORTn using a best guess here.
-                portSelection = get("PORT" + ordinal, PortSource.FROM_PORT_ORDINAL);
+    private PortSelection get(String springPropertyName, String namedPort) {
+        PortSelection portSelection = get(springPropertyName, springPropertyName, PortSource.FROM_SPRING_PROPERTY);
```
Comment (dkaukov) [inline otj-server-core/src/main/java/com/opentable/server/PortSelector.java:105]: We need to check is property supplied first

File: otj-server-core/src/main/java/com/opentable/server/PortSelector.java [hunk 3]
```diff
@@ -92,80 +101,62 @@ public PortSelector(Environment environment) {
...
+        if (!portSelection.hasValue() || "-1".equals(portSelection.getPayload())) {
+            if (isKubernetes(environment)) {
```
Comment (dkaukov) [inline otj-server-core/src/main/java/com/opentable/server/PortSelector.java:107]: Completely separate between k8s and Singularity

File: otj-server-core/src/main/java/com/opentable/server/EmbeddedJettyBase.java [hunk 4]
```diff
@@ -296,17 +294,17 @@ private ConnectorInfo createConnector(Server server, String name, IntSupplier po
...
         server.addConnector(connector);
         return new ServerConnectorInfo(name, connector, config);
     }
 
-    private int selectPort(IntSupplier nextAssignedPort, ServerConnectorConfig connectorConfig) {
+    private int selectPort(ServerConnectorConfig connectorConfig) {
         int configuredPort = connectorConfig.getPort();
         if (configuredPort < 0) {
-            return nextAssignedPort.getAsInt();
+            return 0;
```
Comment (mikebell90) [inline otj-server-core/src/main/java/com/opentable/server/EmbeddedJettyBase.java:307]: I don't quite understand this change.

File: otj-server-core/src/main/java/com/opentable/server/PortSelector.java [hunk 5]
```diff
@@ -40,19 +44,24 @@
         NOT_FOUND
     }
 
+    private final Environment environment;
+    private int portIndex = 0;
```
Comment (mikebell90) [inline otj-server-core/src/main/java/com/opentable/server/PortSelector.java:48]: should this be an atomic integer. Guess I'll see

File: otj-server-core/src/main/java/com/opentable/server/PortSelector.java (hunk 2 above)
Comment (mikebell90) [reply otj-server-core/src/main/java/com/opentable/server/PortSelector.java:105]: But that goes against the NAMED PORT first, the core port of this PR

File: otj-server-core/src/main/java/com/opentable/server/PortSelector.java [hunk 6]
```diff
@@ -92,80 +101,62 @@ public PortSelector(Environment environment) {
...
-        return PortSelection.empty();
+        return PortSelection.empty(name);
     }
 
-    public PortSelection getWithDefault(final String springProperty, final int ordinal, final String namedPort, int defaultV) {
-        final PortSelection portSelection = get(springProperty, ordinal, namedPort );
-        return portSelection.hasValue() ? portSelection : new PortSelection(String.valueOf(defaultV), PortSource.FROM_DEFAULT_VALUE, String.valueOf(defaultV));
+    public PortSelection getWithDefault(final String springProperty, final String namedPort, int defaultV) {
+        final PortSelection portSelection = get(springProperty, namedPort);
+        return portSelection.hasValue() ? portSelection : new PortSelection(springProperty, String.valueOf(defaultV), PortSource.FROM_DEFAULT_VALUE, String.valueOf(defaultV));
```
Comment (mikebell90) [inline otj-server-core/src/main/java/com/opentable/server/PortSelector.java:133]: Looks wrong. I don't understand why springProperty is in here at all.

File: otj-server-core/src/test/java/com/opentable/server/JettyJmxExportTest.java [hunk 7]
```diff
@@ -42,6 +42,7 @@
 @ActiveProfiles(profiles = "deployed")
 @TestPropertySource(properties = {
         "PORT_JMX=9999",
+        "IS_KUBERNETES=TRUE",
```
Comment (mikebell90) [inline otj-server-core/src/test/java/com/opentable/server/JettyJmxExportTest.java:45]: Why is this needed?

File: otj-server-core/src/main/java/com/opentable/server/EmbeddedJettyBase.java (hunk 4 above)
Comment (mikebell90) [reply otj-server-core/src/main/java/com/opentable/server/EmbeddedJettyBase.java:307]: As in won't this break? It's always assuming its PORT0 in this case.

File: otj-server-core/src/main/java/com/opentable/server/PortSelector.java (hunk 5 above)
Comment (mikebell90) [reply otj-server-core/src/main/java/com/opentable/server/PortSelector.java:48]: Looks like it technically should be, as there's no guarantee of same thread ALWAYS

File: otj-server-core/src/main/java/com/opentable/server/PortSelector.java (hunk 3 above)
Comment (mikebell90) [reply otj-server-core/src/main/java/com/opentable/server/PortSelector.java:107]: Yes, but from what I can tell the named port doesn't magically get used primarily. That will make converting old code hard.

File: otj-server-core/src/main/java/com/opentable/server/PortSelector.java [hunk 8]
```diff
@@ -92,80 +101,62 @@ public PortSelector(Environment environment) {
...
-            // Then the "official" spring property
-            portSelection =  get(springPropertyName, PortSource.FROM_SPRING_PROPERTY);
-            if (!portSelection.hasValue()) {
-                // Finally the PORTn using a best guess here.
-                portSelection = get("PORT" + ordinal, PortSource.FROM_PORT_ORDINAL);
+    private PortSelection get(String springPropertyName, String namedPort) {
+        PortSelection portSelection = get(springPropertyName, springPropertyName, PortSource.FROM_SPRING_PROPERTY);
+        if (!portSelection.hasValue() || "-1".equals(portSelection.getPayload())) {
+            if (isKubernetes(environment)) {
+                portSelection = get(springPropertyName, namedPort, PortSource.FROM_PORT_NAMED);
```
Comment (mikebell90) [inline otj-server-core/src/main/java/com/opentable/server/PortSelector.java:108]: Now this can return no value, there's no fall back.

File: otj-server-core/src/main/java/com/opentable/server/PortSelector.java [hunk 9]
```diff
@@ -92,80 +101,62 @@ public PortSelector(Environment environment) {
...
+            } else {
+                portSelection = get(springPropertyName, "PORT" + portIndex, PortSource.FROM_PORT_ORDINAL);
+                portIndex++;
```
Comment (mikebell90) [inline otj-server-core/src/main/java/com/opentable/server/PortSelector.java:111]: This logic is only true for http connectors not jmx or actuator. I think this is very confusing and error prone.

File: otj-server-core/src/main/java/com/opentable/server/PortSelector.java (hunk 1 above)
Comment (mikebell90) [reply otj-server-core/src/main/java/com/opentable/server/PortSelector.java:145]: trim whitespace. Whitespace is mean.
